### PR TITLE
2578: DDBasic V2.0 Theme documentation

### PR DESCRIPTION
**Pull request by martinb14**
http://platform.dandigbib.org/issues/2578

### CODE REVIEW

File: themes/ddbasic/gulpfile.js [hunk 1]
```diff
@@ -82,6 +82,28 @@ gulp.task('sass', 'Process SCSS using libsass',
 );
 
 /**
+ * Usage:
+ * gulp sassprod
+ *
+ * Precompile all sass files into one single bundle.css file in the sass_css folder
+ *
+ */
+gulp.task('sassprod', 'Process SCSS into one css file using libsass',
```
Comment (kasperg) [inline themes/ddbasic/gulpfile.js]: The naming of this task `sassprod` is odd here.

1. As far as I can tell the resulting `bundle.css` is not used for production purposes in the rest of the codebase.
2. If it is to be used for production purposes I would have expected `outputStyle: 'compact'`
3. If it is only used by KSS I would expected it to be called by the `kss` task. On the other hand it actually seems to assemble `bundle.css` on its own.

Perhaps we can delete the task entirely?

File: themes/ddbasic/README.md [hunk 2]
```diff
@@ -1,18 +1,206 @@
...
+Hiding a few fields with CSS is preferred to increasing the number of view modes. 
+
+Rendered view modes is the preferred method of presentation of entities as opposed to fielded views. 
+
+**Rendering ting materials**
+
+A recurring frontend task is to present a list of materials. The prefered procedure is to render an array of material ids as a carousel:
+
+```php
+theme(‘ting_carousel’, array( ‘items’ => array(id1,id2,id3) ) );
```
Comment (kasperg) [inline themes/ddbasic/README.md]: This is not valid code. Note the `‘’` instead of `''`. There should also be a space after each comma.

File: themes/ddbasic/README.md [hunk 3]
```diff
@@ -1,18 +1,206 @@
...
+* full
+* teaser
+* teaser_highlight
+* search_result
+
+ting_object:
+
+* full
+* teaser
+* search_result
```
Comment (kasperg) [inline themes/ddbasic/README.md]: These are actual code elements. I would like to see them wrapped in markers to be formatted properly.

File: themes/ddbasic/README.md [hunk 4]
```diff
@@ -1,18 +1,206 @@
...
+The grid system is a 12 column grid based on [Bourbon Neat](http://neat.bourbon.io/) which is a light weight semantic framework for grids. See [grid settings (temp link)](https://github.com/b14/ding2/blob/ddbasic/themes/ddbasic/sass/configuration/_grid-settings.scss) for details.
+
+A number of [practical breakpoints (temp link)](https://github.com/b14/ding2/blob/ddbasic/themes/ddbasic/sass/configuration/_grid-settings.scss) divides the styling into different device cases based on only one dimension, the screen width. Frontend development in contributed modules should use the established breakpoints, and tests should cover all breakpoint cases.
+
+Specific Drupal / DDBasic rule of thumbs:
+
+* Drupal structures are replicated as nested SCSS objects. 
+* Viewmodes are styled inside the parent entity object
+* Responsive specifications are styled inside the selector they override using the  @include media(..) mixin.
+* Specifications of viewmode rendering in views are placed inside entities using the & - parent selector keeping styling of entities in same file.
```
Comment (kasperg) [inline themes/ddbasic/README.md]: A few more code elements which needs formatting on these two lines.

File: themes/ddbasic/gulpfile.js (hunk 1 above)
Comment (martinb14) [reply themes/ddbasic/gulpfile.js]: To keep styling and documentation apart the tasks are seperate. I have clarified in the comments and changed the gulp task name. Let me know if you prefer to unify the tasks.

File: themes/ddbasic/README.md [hunk 5]
```diff
@@ -1,18 +1,206 @@
...
+* Design for existing manual crop aspect ratios
+* If no aspect matches a desired manual crop, consider using a manual crop very close to the aspect and crop in a derived style or in CSS
+
+## HTML markup
+
+Along with preprocessing done in template.php the [template files (temp link)](https://github.com/b14/ding2ddbasic/tree/master/themes/ddbasic/templates) are the html implementation of the Drupal objects. 
+
+DDBasic follows [Drupal best practices](https://www.drupal.org/coding-standards/css/architecture) for frontend development and Drupals take on CSS and best practices for SCSS.
+
+DDBasic introduces reusable html classes of special purpose which are documented in KSS along with their corresponding SCSS implementation. [See KSS Styleguide (temp link)](http://msd.ding2.server003.b14cms.dk/profiles/ding2/themes/ddbasic/styleguide)
```
[the remark this answers is not in the document]
Comment (martinb14) [reply themes/ddbasic/README.md]: Agree. Let's have a seperate issue for compiled doc pages and related discussions re CI and compliance.
Let me know if core or DDB have a temporary URL instead of the msd URLs for this PR.

File: themes/ddbasic/README.md [hunk 6]
```diff
@@ -1,18 +1,210 @@
...
+~$ gulp sass
+```
+
+### Compile KSS from SCSS
+
+KSS target folder is [sites/all/themes/ddbasic/styleguide (temp link)](http://msd.ding2.server003.b14cms.dk/profiles/ding2/themes/ddbasic/styleguide)
+
+```
+# Compile all sass into one css file
+~$ gulp sassprod
```
Comment (kasperg) [inline themes/ddbasic/README.md]: This task no longer exists.

File: themes/ddbasic/README.md [hunk 7]
```diff
@@ -1,18 +1,210 @@
...
+
+### SCSS lint
+
+The [lint settings](https://github.com/ding2/ding2/tree/master/themes/ddbasic/package.json
+) extends the [stylelint-config-standard](https://github.com/stylelint/stylelint-config-standard/blob/master/index.js) with a few exceptions towards flexibility.
+
+```
+~$ gulp validate-sass 
+```
+Pull requests are required to pass the linter.
```
Comment (kasperg) [inline themes/ddbasic/README.md]: As far as I can tell the current version of the DDBasic theme does not pass this linter.

File: themes/ddbasic/gulpfile.js [hunk 8]
```diff
@@ -81,13 +81,38 @@ gulp.task('sass', 'Process SCSS using libsass',
...
+ * This all in one file is used by kss for the style sheet.
+ *
+ */
+gulp.task('ksssass', 'Process SCSS into one css file using libsass',
+  function () {
+    gulp.src(sassPath)
+      .pipe(sass({outputStyle: 'compact'})
+        .on('error', sass.logError))
+      // Save verbose output for testing purposes.
+      .pipe(gulp.dest('./sass_css_verbose_output'))
```
Comment (kasperg) [inline themes/ddbasic/gulpfile.js]: The verbose output does not seem to be related to KSS. If that is the case then it should not be a part of the task.

File: themes/ddbasic/gulpfile.js [hunk 9]
```diff
@@ -81,13 +81,38 @@ gulp.task('sass', 'Process SCSS using libsass',
...
+      // Minify the css and save it.
+      .pipe(cleanCSS())
+      .pipe(concat('bundle.css'))
+      .pipe(gulp.dest('./sass_css'));
```
Comment (kasperg) [inline themes/ddbasic/gulpfile.js]: This code to generate `bundle.css` seems to be the same as L126-129. Are they both needed?

File: themes/ddbasic/gulpfile.js (hunk 1 above)
Comment (kasperg) [reply themes/ddbasic/gulpfile.js]: I would prefer that the tasks are unified. The result of `ksssass` is not used anywhere else.

File: themes/ddbasic/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasperg) [reply themes/ddbasic/README.md]: :+1: [I created an issue for this](https://platform.dandigbib.org/issues/3183).

We will have to let links be until that is fixed.

File: themes/ddbasic/README.md (hunk 6 above)
Comment (martinb14) [reply themes/ddbasic/README.md]: Fixed

File: themes/ddbasic/gulpfile.js (hunk 9 above)
Comment (martinb14) [reply themes/ddbasic/gulpfile.js]: The two tasks have been joined. ksssass is now deleted.

File: themes/ddbasic/README.md (hunk 7 above)
Comment (martinb14) [reply themes/ddbasic/README.md]: If current stying fails linting, let's push this into a seperate issue.
I'm unsure about how the settings in the gulp file are pushed into CI, if that is what you mean.

File: themes/ddbasic/gulpfile.js (hunk 8 above)
Comment (martinb14) [reply themes/ddbasic/gulpfile.js]: Removed.
